importlab/resolve.py

```diff
@@ -26,11 +26,44 @@ def __init__(self, module_name):
         self.module_name = module_name
 
 
-class ModuleAndDependencies(object):
-    def __init__(self, filename):
-        self.filename = filename
-        self.name, self.ext = os.path.splitext(os.path.basename(self.filename))
-        self.directory = os.path.dirname(self.filename)
+class ResolvedFile(object):
```

Comment:
I'm skeptical of the usefulness of distinguishing between Builtin, Relative, etc. At least for pytype, the two pieces of information I'd want are the path to the file and the name of the thing being imported from it:
(/usr/local/google/home/rechen/requests/requests/compat.py, .compat.bytes),
(/usr/lib/python3.6/zipfile.py, zipfile),
etc.
Knowing that one is a relative import and the other a system import doesn't add much value.

Reply:
the use case i had in mind is for visualisation tools built on top of importlab; it could be useful to know what is a python builtin, what is a pip-installed project and what just came from pythonpath. (Relative is a hack and will be gotten rid of at some point; it's mostly because we resolve a relative import based on the full path to the importing file)

Reply:
True, I hadn't thought of multiple imports. I suggested the name of the imported object because that seemed easy to get from ImportStatement, but all pytype needs to know is where in the file path the module name starts.

Re visualization tools: My concern is that in trying to surface all possibly useful information, we're making it hard to get to the basic information. For pytype, we want to be able to do something like:
```
for file_node in graph.sorted_source_files():
  if file_node.path in pythonpath:
    // this is already implemented
  else:
    module_name = get_module_name(file_node)
    create_default_pyi(get_module_name(file_node))
```
Ideally, `get_module_name` would be as simple as accessing the appropriate attribute on the file node. Currently, it looks like we'd have to do a switch statement on all five subclasses of ResolvedFile and figure out how to reconstruct the module name for each one.

Reply:
good point, i'll add the module name as a field to ResolvedFile, which should take care of the basic case.

Reply:
done

Reply:
I tested this out by changing output.format_file_node to print `f.path, f.module_name, type(f)` and then running importlab --tree requests/requests/adapters.py, and there were some surprising results. For example, one entry was:
```
('/usr/lib/python3/dist-packages/urllib3/contrib/socks.py', 'urllib3.contrib.socks.SOCKSProxyManager', <class 'importlab.resolve.System'>)
```
and when I looked at socks.py, it contained:
```
class SOCKSProxyManager(PoolManager): ...
```
so it looks like SOCKSProxyManager is a class, not a module, and shouldn't appear in the module name.

Reply:
thanks, fixed

Reply:
Now I'm getting an IndexError, when trying to run importlab installed under Python 3.6:

$ importlab requests/requests/adapters.py 
/usr/lib/python3/dist-packages/urllib3/contrib/socks.py:37: DependencyWarning: SOCKS support in urllib3 requires the installation of optional dependencies: specifically, PySocks.  For more information, see https://urllib3.readthedocs.io/en/latest/contrib.html#socks-proxies
  DependencyWarning
Traceback (most recent call last):
  File "/usr/local/bin/importlab", line 4, in <module>
    __import__('pkg_resources').run_script('importlab==0.1.1', 'importlab')
  File "/usr/lib/python3/dist-packages/pkg_resources/__init__.py", line 739, in run_script
    self.require(requires)[0].run_script(script_name, ns)
  File "/usr/lib/python3/dist-packages/pkg_resources/__init__.py", line 1500, in run_script
    exec(code, namespace, namespace)
  File "/usr/local/lib/python3.6/dist-packages/importlab-0.1.1-py3.6.egg/EGG-INFO/scripts/importlab", line 69, in <module>
    sys.exit(main())
  File "/usr/local/lib/python3.6/dist-packages/importlab-0.1.1-py3.6.egg/EGG-INFO/scripts/importlab", line 53, in main
    import_graph = graph.ImportGraph.create(env, args.filenames)
  File "/usr/local/lib/python3.6/dist-packages/importlab-0.1.1-py3.6.egg/importlab/graph.py", line 247, in create
    import_graph.add_file_recursive(os.path.abspath(filename))
  File "/usr/local/lib/python3.6/dist-packages/importlab-0.1.1-py3.6.egg/importlab/graph.py", line 121, in add_file_recursive
    deps, broken, provenance = self.get_file_deps(filename)
  File "/usr/local/lib/python3.6/dist-packages/importlab-0.1.1-py3.6.egg/importlab/graph.py", line 258, in get_file_deps
    f = r.resolve_import(imp)
  File "/usr/local/lib/python3.6/dist-packages/importlab-0.1.1-py3.6.egg/importlab/resolve.py", line 167, in resolve_import
    prefix.endswith(components[-2])):
IndexError: list index out of range

Reply:
sorry, missed a test case (this is a bug caused by treating "from x import *" as a "from" import as well as a "star" import; we should probably refactor to get rid of that, but i'll add a check in the meantime)

Reply:
fixed

Reply:
I'm still seeing some funky module names, e.g.,:
```('/usr/lib/python3.6/_threading_local.py', '_threading_local.local', <class 'importlab.resolve.System'>)```
where `local` is a class in _threading_local.py.

FYI, there are also some flake8 errors in this file. (I don't see any way to add a file-level comment, only line comments and review summaries.)

Reply:
fixed.